test/extended/baremetal/helper.go

```diff
@@ -0,0 +1,198 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+	"sigs.k8s.io/yaml"
+)
+
+// BaremetalTestHelper is an helper class for the baremetal tests,
+// providing support for the most common operations perfomed by
+// the tests. It will also help in reducing the noise in the test
+// definition by hiding technical details
+type BaremetalTestHelper struct {
+	clientSet *kubernetes.Clientset
+	bmcClient dynamic.ResourceInterface
+
+	hostsCreated []*unstructured.Unstructured
+}
+
+// NewBaremetalTestHelper creates a new test helper instance. It is
+// meant to be used in the BeforeEach method
+func NewBaremetalTestHelper(oc *exutil.CLI) *BaremetalTestHelper {
+	clientSet, err := e2e.LoadClientset()
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	return &BaremetalTestHelper{
+		clientSet: clientSet,
+		bmcClient: baremetalClient(oc.AdminDynamicClient()),
+	}
+}
+
+// extraWorkersRetrieveData fetches the information stored in the `extraworkers-secret` previously
+// created via dev-scripts. The secret contains baremetal host and secret definition for every
+// extra worker allocated by dev-scripts
+func (b *BaremetalTestHelper) extraWorkersRetrieveData() (*v1.Secret, error) {
+	ew, err := b.clientSet.CoreV1().Secrets("openshift-machine-api").Get(context.TODO(), "extraworkers-secret", metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return ew, nil
+}
+
+func (b *BaremetalTestHelper) extraWorkersSecretKey(index int) string {
+	return fmt.Sprintf("extraworker-%d-secret", index)
+}
+
+func (b *BaremetalTestHelper) extraWorkerKey(index int) string {
+	return fmt.Sprintf("extraworker-%d-bmh", index)
+}
+
+// getExtraWorkerSecretData gets and decodes the secret associated for the
+// specified extra worker. If not found, the test will fail
+func (b *BaremetalTestHelper) getExtraWorkerSecretData(index int) *v1.Secret {
+	ew, err := b.extraWorkersRetrieveData()
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	yamlData, ok := ew.Data[b.extraWorkersSecretKey(index)]
+	o.Expect(ok).To(o.BeTrue(), fmt.Sprintf("unable to find data for extra worker %d", index))
+
+	var secret v1.Secret
+	err = yaml.Unmarshal(yamlData, &secret)
+	o.Expect(err).ToNot(o.HaveOccurred())
+	return &secret
+}
+
+// getExtraWorkerSecretData gets and decodes the baremetal host associated for the
+// specified extra worker. If not found, the test will fail
+func (b *BaremetalTestHelper) getExtraWorkerData(index int) *unstructured.Unstructured {
+	ew, err := b.extraWorkersRetrieveData()
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	yamlData, ok := ew.Data[b.extraWorkerKey(index)]
+	o.Expect(ok).To(o.BeTrue(), fmt.Sprintf("unable to find data for extra worker %d", index))
+
+	jsonData, err := yaml.YAMLToJSON(yamlData)
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	var host unstructured.Unstructured
+	err = host.UnmarshalJSON(jsonData)
```

Comment:
Isn't there UnmarshallYaml function and that's why we are using `YAMLToJSON`?

Reply:
Yes